File: schema/Sources/WorkoutDBSchema/HealthArchive.swift
import Foundation

public struct HealthArchiveValue: Codable, Sendable, Equatable {
    public let kind: Kind
    public let quantityValue: Double?
    public let unit: String?
    public let categoryValue: Int?
    public let workoutActivityType: String?
    public let durationSeconds: Double?
    public let totalEnergyKcal: Double?
    public let text: String?
    public let reason: String?

    public enum Kind: String, Codable, Sendable {
        case quantity
        case category
        case workout
        case text
        case unsupported
    }

    enum CodingKeys: String, CodingKey {
        case kind
        case quantityValue = "quantity_value"
        case unit
        case categoryValue = "category_value"
        case workoutActivityType = "workout_activity_type"
        case durationSeconds = "duration_seconds"
        case totalEnergyKcal = "total_energy_kcal"
        case text
        case reason
    }

    public init(
        kind: Kind,
        quantityValue: Double? = nil,
        unit: String? = nil,
        categoryValue: Int? = nil,
        workoutActivityType: String? = nil,
        durationSeconds: Double? = nil,
        totalEnergyKcal: Double? = nil,
        text: String? = nil,
        reason: String? = nil
    ) {
        self.kind = kind
        self.quantityValue = quantityValue
        self.unit = unit
        self.categoryValue = categoryValue
        self.workoutActivityType = workoutActivityType
        self.durationSeconds = durationSeconds
        self.totalEnergyKcal = totalEnergyKcal
        self.text = text
        self.reason = reason
    }
}

public struct HealthArchiveRecord: Codable, Sendable, Equatable {
    public let id: String
    public let externalId: String
    public let descriptorId: String
    public let sampleKind: String
    public let sourceBundleIdentifier: String?
    public let startAt: Date?
    public let endAt: Date?
    public let value: HealthArchiveValue
    public let metadata: [String: String]

    enum CodingKeys: String, CodingKey {
        case id
        case externalId = "external_id"
        case descriptorId = "descriptor_id"
        case sampleKind = "sample_kind"
        case sourceBundleIdentifier = "source_bundle_identifier"
        case startAt = "start_at"
        case endAt = "end_at"
        case value
        case metadata
    }

    public init(
        id: String,
        externalId: String,
        descriptorId: String,
        sampleKind: String,
        sourceBundleIdentifier: String? = nil,
        startAt: Date? = nil,
        endAt: Date? = nil,
        value: HealthArchiveValue,
        metadata: [String: String] = [:]
    ) {
        self.id = id
        self.externalId = externalId
        self.descriptorId = descriptorId
        self.sampleKind = sampleKind
        self.sourceBundleIdentifier = sourceBundleIdentifier
        self.startAt = startAt
        self.endAt = endAt
        self.value = value
        self.metadata = metadata
    }
}

public struct HealthArchiveTombstone: Codable, Sendable, Equatable {
    public let id: String
    public let descriptorId: String
    public let externalId: String
    public let observedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case descriptorId = "descriptor_id"
        case externalId = "external_id"
        case observedAt = "observed_at"
    }

    public init(id: String, descriptorId: String, externalId: String, observedAt: Date) {
        self.id = id
        self.descriptorId = descriptorId
        self.externalId = externalId
        self.observedAt = observedAt
    }
}

public struct HealthArchiveUploadPayload: Codable, Sendable, Equatable {
    public let requestSetKey: String
    public let serverNamespace: String
    public let descriptorFingerprint: String
    public let nextCursor: String?
    public let records: [HealthArchiveRecord]
    public let tombstones: [HealthArchiveTombstone]

    enum CodingKeys: String, CodingKey {
        case requestSetKey = "request_set_key"
        case serverNamespace = "server_namespace"
        case descriptorFingerprint = "descriptor_fingerprint"
        case nextCursor = "next_cursor"
        case records
        case tombstones
    }

    public init(
        requestSetKey: String,
        serverNamespace: String,
        descriptorFingerprint: String,
        nextCursor: String?,
        records: [HealthArchiveRecord],
        tombstones: [HealthArchiveTombstone]
    ) {
        self.requestSetKey = requestSetKey
        self.serverNamespace = serverNamespace
        self.descriptorFingerprint = descriptorFingerprint
        self.nextCursor = nextCursor
        self.records = records
        self.tombstones = tombstones
    }
}

public struct HealthArchiveUploadResponse: Codable, Sendable, Equatable {
    public let requestSetKey: String
    public let acknowledgedCursor: String?
    public let recordsReceived: Int
    public let tombstonesReceived: Int
    public let serverTime: Date

    enum CodingKeys: String, CodingKey {
        case requestSetKey = "request_set_key"
        case acknowledgedCursor = "acknowledged_cursor"
        case recordsReceived = "records_received"
        case tombstonesReceived = "tombstones_received"
        case serverTime = "server_time"
    }
}

File: server/workoutdb_server/api/health_archive.py
"""HealthKit personal archive ingestion endpoint."""

from __future__ import annotations

import json
from datetime import UTC, datetime

from fastapi import APIRouter
from sqlalchemy import select

from workoutdb_server.api.deps import CurrentUserId, DbSession
from workoutdb_server.api.schemas import HealthArchiveUploadIn, HealthArchiveUploadOut
from workoutdb_server.models import (
    HealthArchiveRecord,
    HealthArchiveRequestSet,
    HealthArchiveTombstone,
)

router = APIRouter(prefix="/api/health", tags=["health-archive"])


@router.post("/archive", response_model=HealthArchiveUploadOut)
def upload_health_archive(
    payload: HealthArchiveUploadIn,
    db: DbSession,
    user_id: CurrentUserId,
) -> HealthArchiveUploadOut:
    now = datetime.now(UTC)
    for record in payload.records:
        _upsert_record(db, user_id, record, now)
    for tombstone in payload.tombstones:
        _upsert_tombstone(db, user_id, tombstone)
    _upsert_request_set(db, user_id, payload, now)
    db.commit()
    return HealthArchiveUploadOut(
        request_set_key=payload.request_set_key,
        acknowledged_cursor=payload.next_cursor,
        records_received=len(payload.records),
        tombstones_received=len(payload.tombstones),
        server_time=now,
    )


def _upsert_record(db: DbSession, user_id: str, record, now: datetime) -> None:
    existing = (
        db.execute(
            select(HealthArchiveRecord)
            .where(HealthArchiveRecord.user_id == user_id)
            .where(HealthArchiveRecord.descriptor_id == record.descriptor_id)
            .where(HealthArchiveRecord.external_id == record.external_id)
        )
        .scalars()
        .first()
    )
    row = existing or HealthArchiveRecord(
        id=record.id,
        user_id=user_id,
        external_id=record.external_id,
        descriptor_id=record.descriptor_id,
        first_seen_at=now,
    )
    if existing is None:
        db.add(row)
    row.sample_kind = record.sample_kind
    row.source_bundle_identifier = record.source_bundle_identifier
    row.start_at = record.start_at
    row.end_at = record.end_at
    row.unit = record.value.unit
    row.value_json = record.value.model_dump_json(exclude_none=True)
    row.metadata_json = json.dumps(record.metadata, sort_keys=True)
    row.last_seen_at = now


def _upsert_tombstone(db: DbSession, user_id: str, tombstone) -> None:
    existing = (
        db.execute(
            select(HealthArchiveTombstone)
            .where(HealthArchiveTombstone.user_id == user_id)
            .where(HealthArchiveTombstone.descriptor_id == tombstone.descriptor_id)
            .where(HealthArchiveTombstone.external_id == tombstone.external_id)
        )
        .scalars()
        .first()
    )
    if existing is None:
        db.add(
            HealthArchiveTombstone(
                id=tombstone.id,
                user_id=user_id,
                descriptor_id=tombstone.descriptor_id,
                external_id=tombstone.external_id,
                observed_at=tombstone.observed_at,
            )
        )
    else:
        existing.observed_at = tombstone.observed_at


def _upsert_request_set(
    db: DbSession,
    user_id: str,
    payload: HealthArchiveUploadIn,
    now: datetime,
) -> None:
    row = (
        db.execute(
            select(HealthArchiveRequestSet)
            .where(HealthArchiveRequestSet.user_id == user_id)
            .where(HealthArchiveRequestSet.request_set_key == payload.request_set_key)
        )
        .scalars()
        .first()
    )
    if row is None:
        row = HealthArchiveRequestSet(
            user_id=user_id,
            request_set_key=payload.request_set_key,
            records_received=0,
            tombstones_received=0,
        )
        db.add(row)
    row.server_namespace = payload.server_namespace
    row.descriptor_fingerprint = payload.descriptor_fingerprint
    row.acknowledged_cursor = payload.next_cursor
    row.records_received += len(payload.records)
    row.tombstones_received += len(payload.tombstones)
    row.last_uploaded_at = now
